### PR TITLE
py_proto_library uses workspace and pkg paths

### PR DESCRIPTION
**Pull request by daicoden**
All python imports currently are added to the runfiles under __main__
which is only the default when no workspace name is provided. This
change supports both empty workspace names, and specified workspace
names.

This also supports python libs not anchored at root

@gnossen If this looks good - we can close #22784 .

There's still one problem left, which is if you use py_proto_library with import_prefix or strip_import_prefix, the _virtual_imports/<filename> strategy does not work because you can't have two python packages with the same name at different locations without a special package initialization.

https://packaging.python.org/guides/packaging-namespace-packages/

Let me know if I should open an issue to track this, but I think what we could do is add a flag to py_proto_library `pkgutil_namespace = True`. This would then write the __init_ file with `__path__ = __import__('pkgutil').extend_path(__path__, __name__)`

@veblush

### CODE REVIEW

File: bazel/python_rules.bzl [hunk 1]
```diff
@@ -166,6 +166,8 @@ def _generate_pb2_grpc_src_impl(context):
 
     imports = []
     if out_dir.import_path:
+        #TODO: I believe this can be deleted, the rule requires a py_proto_library, and that rule will
```
Comment (gnossen) [inline bazel/python_rules.bzl:169]: I believe you're right. You can go ahead and remove this conditional.

**Comment by awbraunstein** [comment]
I just filed the following issue: https://github.com/grpc/grpc/issues/23580 which I believe is the same as the underlying issue this PR is aiming to fix. My example failure PR https://github.com/grpc/grpc/issues/23580 may have another failing test case that would be worth adding to this PR to ensure it is covered.

Also, if this is not going to be merged, I'd like to know so I can open a separate PR to fix this issue.

**Comment by gnossen** [comment]
@daicoden Sorry for the long wait on this one. I dropped the ball there. There's just [one thing blocking merge](https://source.cloud.google.com/results/invocations/b9b28260-9840-4d4c-b91b-bdf5b1287eb7/log) -- all of the new files need a copyright header comment. Would you mind updating?

**Comment by daicoden** [comment]
Updated @gnossen. Do I need to wory about github/grpc/toplevel_run_tests_invocations/run_tests_sanity_linux_dbg_native, or grpc/core/pull_request/linux/grpc_sanity?

@awbraunstein - Double check, I believe your failure is the same as `no_import_no_strip_test`. Let me know if there is another case.

**Comment by gnossen** [comment]
The sanity tests were failing because of the copyright problem. Rerunning the tests now. Your most recent commit looks like it should fix things up. Thanks!

**Pull request merged**